indexing-service/src/main/java/org/apache/druid/indexing/overlord/TaskQueue.java

```diff
@@ -156,314 +153,248 @@ public TaskQueue(
-                                          + "See overlord logs for more details.";
-            }
-            notifyStatus(task, TaskStatus.failure(task.getId(), errorMessage), errorMessage);
-            continue;
-          }
-          if (taskIsReady) {
-            log.info("Asking taskRunner to run: %s", task.getId());
-            runnerTaskFuture = taskRunner.run(task);
-          } else {
-            // Task.isReady() can internally lock intervals or segments.
-            // We should release them if the task is not ready.
-            taskLockbox.unlockAll(task);
-            continue;
-          }
-        }
-        taskFutures.put(task.getId(), attachCallbacks(task, runnerTaskFuture));
-      } else if (isTaskPending(task)) {
-        // if the taskFutures contain this task and this task is pending, also let the taskRunner
-        // to run it to guarantee it will be assigned to run
-        // see https://github.com/apache/druid/pull/6991
-        taskRunner.run(task);
+  private boolean isTaskReady(Task task)
+  {
+    try {
+      return task.isReady(taskActionClientFactory.create(task));
+    }
+    catch (Exception e) {
+      log.warn(e, "Error while checking if task [%s] is ready to run.", task.getId());
+      final String errorMessage;
+      if (e instanceof MaxAllowedLocksExceededException) {
+        errorMessage = e.getMessage();
+      } else {
+        errorMessage = "Failed while waiting for the task to be ready to run. "
+                       + "See overlord logs for more details.";
       }
+      notifyStatus(task, TaskStatus.failure(task.getId(), errorMessage), errorMessage);
+      return false;
     }
   }
 
-  @VisibleForTesting
-  private void manageInternalPostCritical(
-      final Set<String> knownTaskIds,
-      final Map<String, ListenableFuture<TaskStatus>> runnerTaskFutures
-  )
+  /**
+   * Kills tasks not present in the set of known tasks.
+   */
+  private void killUnknownTasks()
   {
-    // Kill tasks that shouldn't be running
-    final Set<String> tasksToKill = Sets.difference(runnerTaskFutures.keySet(), knownTaskIds);
-    if (!tasksToKill.isEmpty()) {
-      log.info("Asking taskRunner to clean up %,d tasks.", tasksToKill.size());
-
-      // On large installations running several thousands of tasks,
-      // concatenating the list of known task ids can be compupationally expensive.
-      final boolean logKnownTaskIds = log.isDebugEnabled();
-      final String reason = logKnownTaskIds
-              ? StringUtils.format("Task is not in knownTaskIds[%s]", knownTaskIds)
-              : "Task is not in knownTaskIds";
-
-      for (final String taskId : tasksToKill) {
-        try {
-          taskRunner.shutdown(taskId, reason);
-        }
-        catch (Exception e) {
-          log.warn(e, "TaskRunner failed to clean up task: %s", taskId);
-        }
+    final Set<String> knownTaskIds = tasks.keySet();
```

Comment:
I'm a little wary of reintroducing this bug: https://github.com/apache/druid/pull/12901, in the case that `tasks` was changed. Can you check over it? It seems like if `tasks` has changed we might have an issue. But - you've covered most of them with `synchronized`. Can you check in `add()` and `removeTasksInternal`, I think these operate on tasks without being synchronized, so it might lead to some race?

Reply:
Nb: earlier PR https://github.com/apache/druid/pull/12099 is maybe relevant here too for background/history.

Reply:
Thanks for calling this out, @jasonk000 !

The issue described in #12901 is basically a race condition between `notifyStatus` and `manageInternalCritical` (renamed in this patch to `runReadyTasks`), where a task that is being shutdown might get relaunched. In #12901, this was solved by using this order of events:
- Mark the task as `recentlyCompleted`
- Update metadata store
- Finish cleanup of in-memory data structures
All `recentlyCompleted` tasks were ignored in `manageInternalCritical`.

In the new set of changes, I have retained this complete behaviour. As soon as a task is marked as `recentlyCompleted`, it will not be touched by `runReadyTasks` or `killUnknownTasks`. The in-memory data-structures (including `recentlyCompleted` itself) are finally cleaned up atomically only when the task shutdown is finished.

I will try to add some tests for this scenario. Please let me know if you think of any other race condition, would be nice to have tests for all of those.

Reply:
I have tried to ensure that all the changes made in the patch #12901 are retained.

> This patch does the following:
> 1. Fixes the race by adding a recentlyCompletedTasks set that prevents
    the main management loop from doing anything with tasks that are
    currently being cleaned up.
    
The set of `recentlyCompletedTaskIds` is still being used for this purpose.
    
> 2. Switches the order of the critical sections in notifyStatus, so
    metadata store updates happen first. This is useful in case of
    server failures: it ensures that if the Overlord fails in the midst
    of notifyStatus, then completed-task statuses are still available in
    ZK or on MMs for the next Overlord. (Those are cleaned up by
    taskRunner.shutdown, which formerly ran first.) This isn't related
    to the race described above, but is fixed opportunistically as part
    of the same patch.

Order of calls is still the same i.e. metadata store update happens first

> 3. Changes the tasks list to a map. Many operations require retrieval
    or removal of individual tasks; those are now O(1) instead of O(N)
    in the number of running tasks.

- The `tasks` map is now a `ConcurrentHashMap` which still performs O(1) get and remove.
- But now we also have an `activeTaskIdQueue` on which we do an O(n) remove. But this is fine because it does not block other operations of the `TaskQueue`. We need a queue(-like) structure here that must be thread-safe, so I ended up using a `LinkedBlockingDeque`. I am not sure of an alternative that would offer better time complexity.
    
> 4. Changes various log messages to use task ID instead of full task
payload, to make the logs more readable.

Retained as is.

Reply:
The changes in #12099 were a lot of help in cleaning up the `TaskQueue` in this patch. 

The major difference from #12099 is that I am now using a synchronization on an `AtomicBoolean managementRequested` instead of a queue to track if management needs to be done. This choice was made in favour of readability and clarification of intent.

I think you had called out this possible approach here:
> I chose a BlockingQueue implementation because it is easy to reason about the submission / poll / offer ordering. Other options would be Semaphore etc.